app/Client.php

```diff
@@ -23,17 +24,21 @@ private function generateKey(): string
 
     public function save()
     {
-        return Storage::put($this->name . DIRECTORY_SEPARATOR . '.key', $this->key);
+        return Storage::put(
+            $this->name . DIRECTORY_SEPARATOR . '.key',
+            Hash::make($this->key),
+            'private',
+        );
     }
 
-    public static function login(string $name, string $key)
+    public static function loginOrFail(string $name, ?string $key)
     {
-        if (!Storage::exists($name . DIRECTORY_SEPARATOR . '.key')) {
-            return;
-        } elseif (Storage::get($name . DIRECTORY_SEPARATOR . '.key') !== $key) {
-            return;
+        $file = File::findOrFail($name, '.key');
+
+        if (Hash::check($key, $file->binary())) {
+            return new self($name, $key);
         }
 
-        return new self($name, $key);
+        return abort(401, 'API key is missing or invalid');
```

Comment:
Default return should be a success, error should occur when condition isn't met. So if !Hash::check -> abort